src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.IList.cs

```diff
@@ -114,6 +116,116 @@ public void RemoveAt(int index)
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="match"/> is <see langword="null"/>.
+        /// </exception>
+        public int RemoveAll(Func<JsonNode?, bool> match)
+        {
+            if (match == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(match));
+            }
+
+            if (_list is not { } list)
+            {
+                return 0;
+            }
+
+            // Algorithm adapted from List<T>.RemoveAll()
+            int size = list.Count;
+            int freeIndex = 0;   // the first free slot in items array
+
+            // Find the first item which needs to be removed.
+            while (freeIndex < size && !match(list[freeIndex])) freeIndex++;
+            if (freeIndex >= size) return 0;
+
+            int current = freeIndex + 1;
+            while (current < size)
+            {
+                // Find the first item which needs to be kept.
+                while (current < size && match(list[current])) current++;
+
+                if (current < size)
+                {
+                    DetachParent(list[freeIndex]);
+
+                    // copy item to the free slot.
+                    list[freeIndex++] = list[current];
+                    list[current] = null; // prevent it from being detached from the parent in the following for-loop
+                    ++current;
+                }
+            }
+
+            // Detach trailing nodes, as they are skipped in the loop above
+            int removedLength = size - freeIndex;
+            for (int i = 0; i < removedLength; i++)
+            {
+                DetachParent(list[freeIndex + i]);
+            }
+
+            list.RemoveRange(freeIndex, removedLength);
+            return removedLength;
+        }
+
+        /// <summary>
+        ///   Removes a range of elements from the <see cref="JsonArray"/>.
+        /// </summary>
+        /// <param name="index">The zero-based starting index of the range of elements to remove.</param>
+        /// <param name="count">The number of elements to remove.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="index"/> or <paramref name="count"/> is less than 0.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="index"/> and <paramref name="count"/> do not denote a valid range of elements in the <see cref="JsonArray"/>.
+        /// </exception>
+        public void RemoveRange(int index, int count)
+        {
+            if (index < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(index));
+            }
+
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(count));
+            }
+
+            if (_list is not { } list)
```

Comment:
Again, this check should be remove since it's not correct. It also suggests there is current no test coverage for performing this operation against a freshly deserialized `JsonArray` instance using `JsonSerializer`.

Reply:
It can also be null when you do `JsonSerializer.Deserialize<JsonArray>("[1,2,3]");` without having probed the individual elements.